db: prototype of separation of values into blob files

sumeerbhola opened the pull request:
The goal is to reduce write amp without significantly sacrificing locality of access.

Preliminary performance numbers are promising (see TestWriteAmpWithBlobs) which achieves a write amp of 6.8 (versus 15.1 with no blob files) with Lbase == L3 and a uniform random write workload.

There are numerous TODOs, which can be addressed when turning this into production code.

Suggested reading order is the sstable package, then the manifest package and version_set.go, and then compaction.go.

For workloads that care less about locality (e.g. MVCCGet), we could tune for further write amp improvements, but it may also increase the number of files to an unacceptable value -- unlike some other DBs, CockroachDB's use of Pebble grows the target file size by a factor of 2 when descending the levels, so reusing blob files written at higher levels results in higher file count.

Informs #112

sumeerbhola committed:
db: prototype of separation of values into blob files

The goal is to reduce write amp without significantly sacrificing
locality of access.

Preliminary performance numbers are promising (see TestWriteAmpWithBlobs)
which achieves a write amp of 6.8 (versus 15.1 with no blob files) with
Lbase == L3 and a uniform random write workload.

There are numerous TODOs, which can be addressed when turning this into
production code.

Suggested reading order is the sstable package, then the manifest package
and version_set.go, and then compaction.go.

For workloads that care less about locality (e.g. MVCCGet),
we could tune for further write amp improvements, but it may also increase
the number of files to an unacceptable value -- unlike some other DBs,
CockroachDB's use of Pebble grows the target file size by a factor of 2
when descending the levels, so reusing blob files written at higher levels
results in higher file count.

Informs #112

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/pebble/2897)
<!-- Reviewable:end -->

sumeerbhola reviewed (commented):
The prototype was begun in Jan and I have not bothered rebasing, hence the conflicts.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/2897#-:-NeG4df36fFBv5xrm3NU:b-2txvr0)* status: 0 of 68 files reviewed, all discussions resolved



<!-- Sent from Reviewable.io -->

sumeerbhola committed:
db: add BenchmarkReadWithBlobs

```
BenchmarkReadWithBlobs/write-blobs=false/read-value=false/has-cache=false-10         	 5000000	       202.7 ns/op	         0.03244 misses/op	       0 B/op	       0 allocs/op
BenchmarkReadWithBlobs/write-blobs=false/read-value=false/has-cache=true-10          	 5000000	       169.5 ns/op	         0 misses/op	       0 B/op	       0 allocs/op
BenchmarkReadWithBlobs/write-blobs=false/read-value=true/has-cache=false-10          	 5000000	       207.1 ns/op	         0.03244 misses/op	       0 B/op	       0 allocs/op
BenchmarkReadWithBlobs/write-blobs=false/read-value=true/has-cache=true-10           	 5000000	       172.4 ns/op	         0 misses/op	       0 B/op	       0 allocs/op

BenchmarkReadWithBlobs/write-blobs=true/read-value=false/has-cache=false-10          	 5000000	        72.07 ns/op	         0.001193 misses/op	       0 B/op	       0 allocs/op
BenchmarkReadWithBlobs/write-blobs=true/read-value=false/has-cache=true-10           	 5000000	        44.26 ns/op	         0 misses/op	       0 B/op	       0 allocs/op
BenchmarkReadWithBlobs/write-blobs=true/read-value=true/has-cache=false-10           	 5000000	       251.5 ns/op	         0.03252 misses/op	       0 B/op	       0 allocs/op
BenchmarkReadWithBlobs/write-blobs=true/read-value=true/has-cache=true-10            	 5000000	       127.6 ns/op	         0.003208 misses/op	       0 B/op	       0 allocs/op
```

sumeerbhola reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/2897)* status: 0 of 68 files reviewed, 1 unresolved discussion

___
*[`db_test.go` line 1576 at r3](https://reviewable.io/reviews/cockroachdb/pebble/2897#-Ng_b_OaCOMxlo8fXzgY:-Ng_b_Ob-TqWrSes7kxa:beixhrt) ([raw file](https://github.com/cockroachdb/pebble/blob/06743e63fdff79cf1a4a8a8c046828353764bfc1/db_test.go#L1576)):*
> ```Go
> BenchmarkReadWithBlobs/write-blobs=true/read-value=true/has-cache=false-10           	 5000000	       251.5 ns/op	         0.03252 misses/op	       0 B/op	       0 allocs/op
> BenchmarkReadWithBlobs/write-blobs=true/read-value=true/has-cache=true-10            	 5000000	       127.6 ns/op	         0.003208 misses/op	       0 B/op	       0 allocs/op
> ```

Read numbers are very promising.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/2897)* status: 0 of 68 files reviewed, 2 unresolved discussions

___
*[`db_test.go` line 1578 at r3](https://reviewable.io/reviews/cockroachdb/pebble/2897#-Ng_ev0Z5WIflmDVy_Ur:-Ng_ev0Z5WIflmDVy_Us:b-exmq1h) ([raw file](https://github.com/cockroachdb/pebble/blob/06743e63fdff79cf1a4a8a8c046828353764bfc1/db_test.go#L1578)):*
> ```Go
> 
> TODO(sumeer): the 127.6 ns/op in the last bench versus its counterpart with 172.4 ns/op needs
> investigation. I didn't expect blobs to be faster.
> ```

There is nothing special about the cpu profiles. With blob files the calls to `findNextEntry` and `nextUserKey` are much faster. It may be a CPU caching effect, since the 1000 byte value is not inline. 


<!-- Sent from Reviewable.io -->

sumeerbhola committed:
additional commentary
